content/FIDO2/Libraries/Using_a_library.adoc

```diff
@@ -0,0 +1,309 @@
+              Optional.empty()
+          )
+      );
+      registerRequestStorage.put(request.getRequestId(), request);
+      return new Right(request);
+  } else {
+      return new Left("The username \"" + username + "\" is already registered.");
+  }
+}
+----
+
+
+[source, java]
+----
+// handles HTTPS requests to /register/finish
+public Either<List<String>, SuccessfulRegistrationResult> finishRegistration(String responseJson) {
+  ...
+  RegistrationResponse response = null;
+  ...
+      response = jsonMapper.readValue(responseJson, RegistrationResponse.class);
+  ...
+
+  RegistrationRequest request = registerRequestStorage.getIfPresent(response.getRequestId());
+  registerRequestStorage.invalidate(response.getRequestId());
+
+  ...
+      Try<RegistrationResult> registrationTry = rp.finishRegistration(
+          request.getPublicKeyCredentialCreationOptions(),
+          response.getCredential(),
+          Optional.empty()
+      );
+
+      if (registrationTry.isSuccess()) {
+          return Right.apply(
+              new SuccessfulRegistrationResult(
+                  request,
+                  response,
+                  addRegistration(
+                      request.getUsername(),
+                      request.getPublicKeyCredentialCreationOptions().user(),
+                      request.getCredentialNickname(),
+                      response,
+                      registrationTry.get()
+                  ),
+                  registrationTry.get().attestationTrusted()
+              )
+          );
+      } else {
+          ...
+          return Left.apply(Arrays.asList("Registration failed!", registrationTry.failed().get().getMessage()));
+      }
+
+  ...
+}
+----
+
+
+==== Authentication ====
+
+[source, java]
+----
+// handles HTTPS requests to /authenticate
+public AssertionRequest startAuthentication(Optional<String> username) {
+    ...
+    AssertionRequest request = new AssertionRequest(
+        username,
+        U2fB64Encoding.encode(challengeGenerator.generateChallenge()),
+        rp.startAssertion(
+            username.map(userStorage::getCredentialIdsForUsername),
+            Optional.empty()
+        )
+    );
+
+    assertRequestStorage.put(request.getRequestId(), request);
+
+    return request;
+}
+----
+
+[source, java]
+----
+// handles HTTPS requests to /authenticate/finish
+public Either<List<String>, SuccessfulAuthenticationResult> finishAuthentication(String responseJson) {
+    ...
+
+    final AssertionResponse response;
+    ...
+        response = jsonMapper.readValue(responseJson, AssertionResponse.class);
+    ...
+
+    AssertionRequest request = assertRequestStorage.getIfPresent(response.getRequestId());
+    assertRequestStorage.invalidate(response.getRequestId());
+
+    ...
+        Optional<String> returnedUserHandle = Optional.ofNullable(response.getCredential().response().userHandleBase64());
+
+        final String username;
```

Comment:
Hm, I can probably move a bunch of this `username`/`userHandle` logic into one of the library components... Let me think about that. At any rate, most of L156-L169 is distracting implementation details that would be nice to get out of the example code.